Load the config file by default on libcrypto initialization

levitte opened the pull request:
As specified in the 3.0.0 design document:

> By default, OpenSSL 3.0 will load a configuration file (which
> contains global properties and other settings) automatically without
> explicit application API calls. This will occur in libcrypto. Note
> that in OpenSSL 1.1.1 the configuration file is automatically loaded
> only by the default (automatic) initialisation of libssl.

Ref: https://www.openssl.org/docs/OpenSSL300Design.html#property-based-algorithmselection

levitte committed:
Load the config file by default on libcrypto initialization

As specified in the 3.0.0 design document:

> By default, OpenSSL 3.0 will load a configuration file (which
> contains global properties and other settings) automatically without
> explicit application API calls. This will occur in libcrypto. Note
> that in OpenSSL 1.1.1 the configuration file is automatically loaded
> only by the default (automatic) initialisation of libssl.

Ref: https://www.openssl.org/docs/OpenSSL300Design.html#property-based-algorithmselection

levitte committed:
fixup! Load the config file by default on libcrypto initialization

levitte committed:
OPENSSL_init_crypto(): rearrange so config file loading happens last

This ensures that all builtin things that should be loaded are loaded
at the time for configuration.

levitte committed:
Eliminate unnecessary and disruptive calls to OPENSSL_init_crypto()

There's no need to initialize all of libcrypto just to:

- do_err_strings_init(): create a lock and a hash table
- do_ex_data_init(): create a lock.
- ossl_init_thread_start(): create local thread storage.
- do_default_context_init(): create a default context.
- do_provider_store_init(): create a provider store context index.

Furthermore:

- ERR_get_state(): there's no need to load the crypto strings here.

All these calls caused lock-ups, since they all eventually called as
part of a OPENSSL_init_crypto() call.

mattcaswell commented on crypto/conf/conf_sap.c:
```diff
@@ -60,10 +60,6 @@ int openssl_config_int(const OPENSSL_INIT_SETTINGS *settings)
-#ifndef OPENSSL_NO_ENGINE
-    /* Need to load ENGINEs */
-    ENGINE_load_builtin_engines();
-#endif
```
Why is this no longer required?

levitte replied:
It's the wrong spot to call `ENGINE_load_builtin_engines()`, as this code is called by `OPENSSL_init_crypto()` and `ENGINE_load_builtin_engines()` also calls `OPENSSL_init_crypto()`.

mattcaswell commented on crypto/err/err.c:
```diff
@@ -742,9 +740,6 @@ ERR_STATE *ERR_get_state(void)
         }
-
-        /* Ignore failures from these */
-        OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL);
```
Why isn't this required any more? (it seems to be the only place in the library where we set this option)

levitte replied:
It's just that ERR functions may be called down the line from `OPENSSL_init_crypto()`, so this causes an init lockup.

levitte replied:
We *might* want to consider adding a `OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL)` in the error printing functions, or the error string retrieval functions.  That's where they are actually needed, and those are called explicitly by the application

mattcaswell replied:
But if you don't do this, I can't see how the crypto strings get loaded.

levitte replied:
Yeahok...  so I gotta ask, since we get them loaded unconditionally here, what's the reason we pretend that it's optional to do so?  In other words, why not this?

``` diff
diff --git a/crypto/init.c b/crypto/init.c
index d8acfa03ec..afbb5a6872 100644
--- a/crypto/init.c
+++ b/crypto/init.c
@@ -615,15 +615,14 @@ int OPENSSL_init_crypto(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings)
     if (!RUN_ONCE(&load_crypto_nodelete, ossl_init_load_crypto_nodelete))
         return 0;
 
-    if ((opts & OPENSSL_INIT_NO_LOAD_CRYPTO_STRINGS)
-            && !RUN_ONCE_ALT(&load_crypto_strings,
-                             ossl_init_no_load_crypto_strings,
-                             ossl_init_load_crypto_strings))
-        return 0;
-
-    if ((opts & OPENSSL_INIT_LOAD_CRYPTO_STRINGS)
-            && !RUN_ONCE(&load_crypto_strings, ossl_init_load_crypto_strings))
+    if ((opts & OPENSSL_INIT_NO_LOAD_CRYPTO_STRINGS)) {
+        if (!RUN_ONCE_ALT(&load_crypto_strings,
+                          ossl_init_no_load_crypto_strings,
+                          ossl_init_load_crypto_strings))
+            return 0;
+    } else if (!RUN_ONCE(&load_crypto_strings, ossl_init_load_crypto_strings)) {
         return 0;
+    }
 
     if ((opts & OPENSSL_INIT_NO_ADD_ALL_CIPHERS)
             && !RUN_ONCE_ALT(&add_all_ciphers, ossl_init_no_add_all_ciphers,
```

levitte replied:
But not again that we could also place `OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL)` in the functions that are used to retrieve strong values when exploring the error stack, i.e. as late as possible.

levitte replied:
I.e. this:

``` diff
diff --git a/crypto/err/err.c b/crypto/err/err.c
index c558bfa62b..d7a0c1ede8 100644
--- a/crypto/err/err.c
+++ b/crypto/err/err.c
@@ -589,6 +589,8 @@ void ERR_error_string_n(unsigned long e, char *buf, size_t len)
     if (len == 0)
         return;
 
+    OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL);
+
     l = ERR_GET_LIB(e);
     ls = ERR_lib_error_string(e);
     if (ls == NULL) {
```

levitte replied:
The more I think about it, the more I prefer the `crypto/init.c` diff...

levitte replied:
In the end, I chose the latter.  That seems like the sane way to avoid lockups.

mattcaswell commented on crypto/err/err.c:
```diff
@@ -293,8 +293,6 @@ static void ERR_STATE_free(ERR_STATE *s)
 DEFINE_RUN_ONCE_STATIC(do_err_strings_init)
 {
-    if (!OPENSSL_init_crypto(0, NULL))
-        return 0;
```
Why is this not needed?

levitte replied:
All we do in this function is to create a lock and a hash table.  There's zero need to initialize the rest of libcrypto here, specifically.  Also, lockup.

mattcaswell commented on crypto/ex_data.c:
```diff
@@ -37,8 +37,6 @@ static CRYPTO_ONCE ex_data_init = CRYPTO_ONCE_STATIC_INIT;
 DEFINE_RUN_ONCE_STATIC(do_ex_data_init)
 {
-    if (!OPENSSL_init_crypto(0, NULL))
-        return 0;
```
Why is this not needed

levitte replied:
Lockup.  Also, just to create a lock?

mattcaswell commented on crypto/init.c:
```diff
@@ -431,9 +431,6 @@ int ossl_init_thread_start(uint64_t opts)
     struct thread_local_inits_st *locals;
 
-    if (!OPENSSL_init_crypto(0, NULL))
-        return 0;
```
Why is this not needed

levitte replied:
Lockup, and really not needed just to create local storage.

mattcaswell commented on crypto/provider_core.c:
```diff
@@ -134,8 +134,7 @@ static const OPENSSL_CTX_METHOD provider_store_method = {
 DEFINE_RUN_ONCE_STATIC(do_provider_store_init)
 {
-    return OPENSSL_init_crypto(0, NULL)
-        && (provider_store_index =
```
Why is this not needed?

levitte replied:
Lockup when the configuration file contains provider data.  Please, not needed just to create a lock.

Frankly, it was added "just 'cause", i.e. there was no real thought about from my end, so...

mattcaswell reviewed (changes requested):
Before taking out all of the OPENSSL_init_crypto calls I think we just need to discuss each one a bit and make sure we can justify it.

levitte committed:
fixup! Load the config file by default on libcrypto initialization

levitte commented on crypto/context.c:
```diff
@@ -38,8 +38,7 @@ static void do_default_context_deinit(void)
-    return OPENSSL_init_crypto(0, NULL)
-        && context_init(&default_context)
+    return context_init(&default_context)
         && OPENSSL_atexit(do_default_context_deinit);
```
I've realised that if `OPENSSL_atexit()` is called, a `OPENSSL_init_crypto(OPENSSL_INIT_BASE_ONLY, NULL)` must be made first...  so instead of removals, I'll just change the first argument

levitte replied:
Wait, that wasn't true, so never mind (OPENSSL_init_crypto(OPENSSL_INIT_BASE_ONLY,NULL)` does *nothing* regarding atexit processing anyway)

levitte committed:
Ensure libcrypto error strings are loaded, but late

Do this when ERR_error_string_n() gets called, so we get the strings
just in time.

levitte committed:
test/shlibloadtest.c: call ERR_error_string_n() instead of ERR_get_error()

This ensures OPENSSL_init_crypto() gets called.

levitte commented:
Closing this one, as the way it's handled here doesn't work with the way the init stuff is constructed.

I've started over in a different branch, PR coming up soonish
